### PR TITLE
ZCS-16181 : Password expiry reminder via. email

### PR DESCRIPTION
**Pull request by viditZim**
https://synacor.atlassian.net/browse/ZCS-16181

Problem: Admin should be able to set password reminders for the users so that they are informed about the upcoming password expiry of their account and accordingly can take the required action.

Solution: 

1.  Added zimbraFeaturePasswordExpiryReminderEnabled LDAP attribute to enable the email reminder.
2. Created Notifier to fetch all the accounts with reminder enabled and whose password is about expire and then send the reminder emails.

### CODE REVIEW

File: store/conf/attrs/zimbra-attrs.xml [hunk 1]
```diff
@@ -10560,4 +10560,9 @@ TODO: delete them permanently from here
   <defaultCOSValue>FALSE</defaultCOSValue>
   <desc>Feature to enable delivery status notification</desc>
 </attr>
+
+<attr id="4136" name="zimbraFeaturePasswordExpiryReminderEnabled" type="boolean" cardinality="single" optionalIn="cos,account" flags="accountInherited,accountInfo" since="10.1.4">
```
Comment (rcyarrapothu) [inline store/conf/attrs/zimbra-attrs.xml:10564]: don't think we can release a new ldap attribute in patch versions

Comment (viditZim) [reply store/conf/attrs/zimbra-attrs.xml:10564]: This is being checked to get released.

Comment (shrutig0510) [reply store/conf/attrs/zimbra-attrs.xml:10564]: @rcyarrapothu Please follow this ticket https://synacor.atlassian.net/browse/ZCS-16186 . Its till under discussion and validation.

File: store/src/java/com/zimbra/cs/service/mail/PasswordExpiryNotifier.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    private static final int PASSWORD_REMINDER_BATCH_SIZE = 500;
+
+    public static void main(String[] args) {
+        Thread thread = new Thread(new PasswordExpiryNotifier());
+        thread.start();
+    }
+
+    protected static void sendReminder() {
+        try {
+            LdapProv ldapProv = LdapProv.getInst();
```
Comment (shrutig0510) [inline store/src/java/com/zimbra/cs/service/mail/PasswordExpiryNotifier.java:68]: Please create separate method to get all eligible accounts from ldap

File: store/src/java/com/zimbra/cs/service/mail/PasswordExpiryNotifier.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+    private static void sendMail(Account account, long deadline, String expiresOn) {
+        try {
+            Session session = JMSession.getSmtpSession(account);
+            MimeMessage mimeMessage = new MimeMessage(session);
+            mimeMessage.setFrom(new InternetAddress("no-reply", "Password Expiration Notification"));
+            mimeMessage.addRecipient(Message.RecipientType.TO, new InternetAddress(account.getAttr(Provisioning.A_zimbraMailDeliveryAddress)));
+            mimeMessage.setSubject("Your password expires in " + deadline + " Days");
+            String userName = account.getAttr(Provisioning.A_cn);
+            String textContent = "Dear " + userName + ",\n" +
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/cs/service/mail/PasswordExpiryNotifier.java:154]: @shrutig0510 do we need multi-lingual support for this email?

File: store/src/java/com/zimbra/cs/service/mail/PasswordExpiryNotifier.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        for (NamedEntry ne : batch) {
+            Account account = (Account) ne;
+            int maxAge = account.getIntAttr(Provisioning.A_zimbraPasswordMaxAge, 0);
+            Date lastChange = account.getGeneralizedTimeAttr(Provisioning.A_zimbraPasswordModifiedTime, null);
+            if (lastChange != null) {
+                long last = lastChange.getTime();
+                long curr = System.currentTimeMillis();
+                long expires = last + (Constants.MILLIS_PER_DAY * maxAge);
+                Date expiryDate = new Date(expires);
+                DateFormat formatter = new SimpleDateFormat("MM/dd/yyyy");
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/cs/service/mail/PasswordExpiryNotifier.java]: Please check for existing constants of date format to use here

File: store/src/java/com/zimbra/cs/service/mail/PasswordExpiryNotifier.java (hunk 2 above)
Comment (viditZim) [reply store/src/java/com/zimbra/cs/service/mail/PasswordExpiryNotifier.java:68]: Done

File: store/src/java/com/zimbra/cs/service/mail/PasswordExpiryNotifier.java (hunk 3 above)
Comment (shrutig0510) [reply store/src/java/com/zimbra/cs/service/mail/PasswordExpiryNotifier.java:154]: Yes.

Comment (viditZim) [reply store/src/java/com/zimbra/cs/service/mail/PasswordExpiryNotifier.java:154]: Hi @dasiyogesh , separate ticket will be created for multi-lingual.

**Review by Gopal-Moolchandani — APPROVED** [review]

File: store/conf/attrs/zimbra-attrs.xml [hunk 5]
```diff
@@ -10560,4 +10560,9 @@ TODO: delete them permanently from here
...
+  <defaultCOSValue>FALSE</defaultCOSValue>
+  <desc>Feature to enable password expiry reminder</desc>
```
Comment (shrutig0510) [inline store/conf/attrs/zimbra-attrs.xml]:
```suggestion
  <desc>Feature to enable/disable password expiry reminder. The default value is FALSE. </desc>
```

File: store/src/java/com/zimbra/cs/service/mail/PasswordExpiryNotifier.java [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+    public static void main(String[] args) {
+        Thread thread = new Thread(new PasswordExpiryNotifier());
+        thread.start();
+    }
+
+    protected static void sendReminder() {
+        try {
+            LdapProv ldapProv = LdapProv.getInst();
+            List<NamedEntry> accounts = fetchEligibleAccounts(ldapProv);
```
Comment (shrutig0510) [inline store/src/java/com/zimbra/cs/service/mail/PasswordExpiryNotifier.java]:
```suggestion
            List<NamedEntry> accounts = findAccountsWithMaxAgePasswordAndReminderEnabled
(ldapProv);
```

File: store/src/java/com/zimbra/cs/service/mail/PasswordExpiryNotifier.java [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+            MimeBodyPart textPart = new MimeBodyPart();
+            textPart.setText(textContent);
+            MimeBodyPart htmlPart = new MimeBodyPart();
+            htmlPart.setContent(htmlContent, "text/html");
+            multipart.addBodyPart(textPart);
+            multipart.addBodyPart(htmlPart);
+            mimeMessage.setContent(multipart);
+            Transport.send(mimeMessage);
+        } catch (Exception e) {
+            ZimbraLog.store.error("Failed to send email", e);
```
Comment (shrutig0510) [inline store/src/java/com/zimbra/cs/service/mail/PasswordExpiryNotifier.java]:
```suggestion
            ZimbraLog.store.error("Failed to send email for account %s ", account.getName(), e);
```

File: store/src/java/com/zimbra/cs/service/mail/PasswordExpiryNotifier.java [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+import javax.mail.internet.MimeMessage;
+import javax.mail.internet.MimeMultipart;
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+
```
Comment (shrutig0510) [inline store/src/java/com/zimbra/cs/service/mail/PasswordExpiryNotifier.java:47]: Please add api docs

Comment (viditZim) [reply store/src/java/com/zimbra/cs/service/mail/PasswordExpiryNotifier.java:47]: Done

**Review by shrutig0510 — APPROVED** [review]

**Pull request merged**